File: likeAPI/like.js
const express = require('express');
const app = express();
const cors = require('cors');
const sqlite3 = require('sqlite3');


app.use(cors());
app.use(express.json());

const db = new sqlite3.Database('../app.db');

// pour add un like
app.post('/like', (request, resultat) => {
  const { user_id, post_id, comment_id } = request.body;

  const stmt = db.prepare(
    `INSERT INTO likes (user_id, post_id, comment_id) VALUES (?, ?, ?)`
  );

  stmt.run([user_id, post_id || null, comment_id || null], function (err) {
    if (err) {
      return resultat.status(500).json({ error: err.message });
    }
    resultat.status(201).json({ id: this.lastID, message: 'like ajouté' });
  });
});

// pour le delete
app.delete('/like', (request, resultat) => {
  const { user_id, post_id, comment_id } = request.body;

  const stmt = db.prepare(
    `DELETE FROM likes WHERE user_id = ? AND post_id IS ? AND comment_id IS ?`
  );

  stmt.run([user_id, post_id || null, comment_id || null], function (err) {
    if (err) {
      return resultat.status(500).json({ error: err.message });
    }
    resultat.status(200).json({ message: 'Like retiré' });
  });
});

// pour le noombre de like
app.get('/like/count', (request, resultat) => {
  const { post_id, comment_id } = request.query;

  const query = `
    SELECT COUNT(*) as count
    FROM likes
    WHERE post_id IS ? AND comment_id IS ?
  `;

  db.get(query, [post_id || null, comment_id || null], (err, row) => {
    if (err) {
      return resultat.status(500).json({ error: err.message });
    }
    resultat.json({ count: row.count });
  });
});

app.get('/', (req, res) => {
  res.send('API LIKE PRETE');
});
app.listen(3001, () => {
  console.log('API Likes SQLite démarrée sur http://localhost:3001');
});

app.get('/like/has-liked', (req, res) => {
  const { user_id, post_id } = req.query;

  db.get(
    `SELECT COUNT(*) as count FROM likes WHERE user_id = ? AND post_id = ?`,
    [user_id, post_id],
    (err, row) => {
      if (err) return res.status(500).json({ error: err.message });
      res.json({ liked: row.count > 0 });
    }
  );
});

File: likeAPI/like.go
package main

import (
	"database/sql"
	"net/http"

	"github.com/gin-gonic/gin"
	_ "github.com/mattn/go-sqlite3"
)

func main() {
	router := gin.Default()

	db, err := sql.Open("sqlite3", "./app.db")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	router.POST("/like", func(c *gin.Context) {
		var body struct {
			UserID int `json:"user_id"`
			PostID int `json:"post_id"`
		}
		if err := c.BindJSON(&body); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid JSON"})
			return
		}

		// verifier si le like existe
		var count int
		err := db.QueryRow("SELECT COUNT(*) FROM likes WHERE user_id=? AND post_id=?", body.UserID, body.PostID).Scan(&count)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "DB error"})
			return
		}

		if count > 0 {
			// Unlike
			_, err := db.Exec("DELETE FROM likes WHERE user_id=? AND post_id=?", body.UserID, body.PostID)
			if err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "DB error"})
				return
			}
			c.JSON(http.StatusOK, gin.H{"liked": false})
		} else {
			// Like
			_, err := db.Exec("INSERT INTO likes (user_id, post_id) VALUES (?, ?)", body.UserID, body.PostID)
			if err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "DB error"})
				return
			}
			c.JSON(http.StatusOK, gin.H{"liked": true})
		}
	})

	router.Run(":8080")
}
